homeassistant/components/calendar/trigger.py

```diff
@@ -0,0 +1,171 @@
+"""Offer calendar automation rules."""
+from __future__ import annotations
+
+import datetime
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components.automation import (
+    AutomationActionType,
+    AutomationTriggerInfo,
+)
+from homeassistant.const import CONF_ENTITY_ID, CONF_EVENT, CONF_PLATFORM
+from homeassistant.core import CALLBACK_TYPE, HassJob, HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.event import (
+    async_track_point_in_utc_time,
+    async_track_time_interval,
+)
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.util import dt as dt_util
+
+from . import DOMAIN, CalendarEventDevice
+
+_LOGGER = logging.getLogger(__name__)
+
+EVENT_START = "start"
+UPDATE_INTERVAL = datetime.timedelta(minutes=15)
+
+TRIGGER_SCHEMA = cv.TRIGGER_BASE_SCHEMA.extend(
+    {
+        vol.Required(CONF_PLATFORM): DOMAIN,
+        vol.Required(CONF_ENTITY_ID): cv.entity_id,
+        vol.Required(CONF_EVENT, default=EVENT_START): vol.In({EVENT_START}),
+    }
+)
+
+
+class CalendarEventListener:
+    """Helper class to listen to calendar events."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        job: HassJob,
+        trigger_data: dict[str, Any],
+        entity: CalendarEventDevice,
+    ) -> None:
+        """Initialize CalendarEventListener."""
+        self._hass = hass
+        self._job = job
+        self._trigger_data = trigger_data
+        self._entity = entity
+        self._unsub_event: CALLBACK_TYPE | None = None
+        self._unsub_refresh: CALLBACK_TYPE | None = None
+        # Upcoming set of events with their trigger time
+        self._events: list[tuple[datetime.datetime, dict[str, Any]]] = []
+
+    async def async_attach(self) -> None:
+        """Attach a calendar event listener."""
+        now = dt_util.utcnow()
+        await self._fetch_events(now)
+        self._unsub_refresh = async_track_time_interval(
+            self._hass, self._handle_refresh, UPDATE_INTERVAL
+        )
+        self._listen_next_calendar_event()
+
+    @callback
+    def async_detach(self) -> None:
+        """Detach the calendar event listener."""
+        assert self._unsub_event is not None
+        assert self._unsub_refresh is not None
+
+        self._unsub_event()
+        self._unsub_event = None
+        self._unsub_refresh()
+        self._unsub_refresh = None
+
+    async def _fetch_events(self, now: datetime.datetime) -> None:
+        """Update the set of eligible events."""
+        start_date = now
+        end_date = now + UPDATE_INTERVAL * 2
+        events = await self._entity.async_get_events(self._hass, start_date, end_date)  # type: ignore[no-untyped-call]
```

Comment:
We shouldn't add type ignores that aren't needed when we'll merge this after the other PR that types this target method.

Reply:
Would you recommend I not send this PR and wait for the other to be resolved next time? Since these were /mostly/ orthogonal i thought it was worth this "hack" that i would cleanup since the other would likely go in fairly quickly. I realize this can cause churn, but took the risk.